libs/api/utils/src/lib/api-utils.ts

```diff
@@ -20,11 +44,61 @@ export async function validateApiKey(apiKey: string): Promise<ApiErrorCode | und
   if (key.is_deleted) return 'PUBLIC_API_KEY_HAS_BEEN_DELETED'
   if (!key.is_active) return 'PUBLIC_API_KEY_IS_NOT_ACTIVE'
   if (isDateBeforeNow(key.expiration_date)) return 'PUBLIC_API_KEY_HAS_BEEN_EXPIRED'
+  API_KEY_TO_ID_CACHE[apiKey] = key.id
+  API_KEY_CONTEXT = apiKey
 
   await incrementKeyUsage(key.id)
   return
 }
 
+export async function resolveApiKeyIdByKey(apiKey: string): Promise<bigint> {
+  const keyId: bigint | undefined = API_KEY_TO_ID_CACHE[apiKey as keyof typeof API_KEY_TO_ID_CACHE]
+  if (keyId) return keyId
+
+  const selectedKey = await prisma.cas__organisation_api_key.findFirstOrThrow({
+    select: { id: true },
+    where: { api_key: apiKey },
+  })
+  API_KEY_TO_ID_CACHE[apiKey] = selectedKey.id
+
+  return selectedKey.id
+}
+
+export async function computeUsage(r: ApiUsageRequest): Promise<ApiUsageReport[]> {
+  const keyIds = await resolveUsageKeyIds(r)
+  const spendRaw = await prisma.$queryRaw<ApiUsageReport[]>`
+   SELECT
+     u.operation_class AS operation,
+     SUM(u.spend) AS spend,
+     COUNT(u.spend) AS operations
+   FROM core__api_key_usage u
+   WHERE u.key_id IN (${Prisma.join(keyIds)})
+   GROUP BY u.operation_class`;
+
+  return spendRaw.map(
+    (record): ApiUsageReport => ({
+      operation: record.operation,
+      spend: Number(record.spend),
+      operations: Number(record.operations),
+    }),
+  )
+}
+
+async function resolveUsageKeyIds(r: ApiUsageRequest): Promise<bigint[]> {
+  if (!('apiKey' in r || 'userId' in r || 'organisationId' in r)) return []
```

Comment:
**thought:** I was thinking whether we shouldn't go this way, but it feels more complex. What do you think?

```typescript
if (!['apiKey', 'userId', 'organisationId'].some(key => key in request)) return []
```

Reply:
Maybe better can be (without key duplicate):

```ts
const where = {
  ...('apiKey' in request ? { api_key: request.apiKey } : {}),
  ...('userId' in request ? { user_id: request.userId } : {}),
  ...('organisationId' in request ? { organisation_id: request.organisationId } : {}),
};
if (!Object.keys(where).length) return []
const keyListRaw = await prisma.cas__organisation_api_key.findMany({
  select: { id: true },
  where
})
```